starknet-client/integration/contract_interaction.rs

```diff
@@ -0,0 +1,176 @@
+async fn test_nullifier_used(test_args: IntegrationTestArgs) -> Result<()> {
+    let mut rng = thread_rng();
+    let client = &test_args.starknet_client;
+
+    // Check that a random nullifier is not used
+    let random_nullifier = Scalar::random(&mut rng);
+    let nullifier_unused = client.check_nullifier_unused(random_nullifier).await?;
+    assert_true_result!(nullifier_unused)?;
+
+    // Call `update_wallet` with a dummy nullifier then check that it is used
+    let dummy_nullifier = Scalar::random(&mut rng);
+    let tx_hash = client
+        .update_wallet(
+            Scalar::one(),
+            dummy_nullifier,
+            None, /* external_transfer */
+            dummy_wallet_share(),
+            dummy_valid_wallet_update_bundle(),
+        )
+        .await?;
+    client.poll_transaction_completed(tx_hash).await?;
+
+    let nullifier_unused = client.check_nullifier_unused(dummy_nullifier).await?;
+    assert_true_result!(!nullifier_unused)
+}
+integration_test_async!(test_nullifier_used);
+
+/// Tests submitting a new wallet and then recovering its shares from the contract
+async fn test_new_wallet(test_args: IntegrationTestArgs) -> Result<()> {
+    let client = &test_args.starknet_client;
+
+    // Deploy a new wallet
+    let (_commitment, public_shares) = deploy_new_wallet(client).await?;
+
+    // Find the transaction that updated the wallet
+    let tx_hash = client
+        .get_public_blinder_tx(public_shares.blinder)
+        .await?
+        .ok_or(eyre::eyre!("No transaction found for public blinder"))?;
+    let recovered_public_shares = client
+        .fetch_public_shares_from_tx(public_shares.blinder, tx_hash)
+        .await?;
+
+    // Check that the recovered public shares are the same as the original ones
+    assert_eq_result!(public_shares, recovered_public_shares)
+}
+integration_test_async!(test_new_wallet);
+
+/// Test submitting a wallet update that succeeds and then recovering its shares
+async fn test_update_wallet(test_args: IntegrationTestArgs) -> Result<()> {
+    let mut rng = thread_rng();
+    let client = &test_args.starknet_client;
+
+    // Deploy a new wallet
+    let (commitment, public_shares) = deploy_new_wallet(client).await?;
+    let nullifier = Scalar::random(&mut rng);
+    let new_shares = dummy_wallet_share();
+
+    // Update the wallet
+    let tx_hash = client
+        .update_wallet(
+            commitment,
+            nullifier,
+            None, /* external_transfer */
+            new_shares.clone(),
+            dummy_valid_wallet_update_bundle(),
+        )
+        .await?;
+    client.poll_transaction_completed(tx_hash).await?;
+
+    // Check that the nullifier is used
+    let nullifier_unused = client.check_nullifier_unused(nullifier).await?;
+    assert_true_result!(!nullifier_unused)?;
+
+    // Find the transaction that updated the wallet
+    let tx_hash = client
+        .get_public_blinder_tx(public_shares.blinder)
```

Comment:
wouldn't the transaction that updated the wallet be indexed by `new_shares.blinder`?
and subsequently, we should check that `recovered_public_shares == new_shares`, right